transport/src/main/java/io/netty/channel/local/LocalChannel.java

```diff
@@ -211,23 +217,19 @@ protected void doClose() throws Exception {
             state = State.CLOSED;
         }
 
-        final LocalChannel peer = this.peer;
         if (peer != null && peer.isActive()) {
-            // Need to execute the close in the correct EventLoop
-            // See https://github.com/netty/netty/issues/1777
-            EventLoop eventLoop = peer.eventLoop();
-
+            // Need to execute the close in the correct EventLoop (see https://github.com/netty/netty/issues/1777).
             // Also check if the registration was not done yet. In this case we submit the close to the EventLoop
-            // to make sure it is run after the registration completes.
-            //
-            // See https://github.com/netty/netty/issues/2144
-            if (eventLoop.inEventLoop() && !registerInProgress) {
-                peer.unsafe().close(unsafe().voidPromise());
+            // to make sure its run after the registration completes (see https://github.com/netty/netty/issues/2144).
+            if (peer.eventLoop().inEventLoop() && !registerInProgress) {
+                doPeerClose(peer, peer.writeInProgress);
             } else {
-                peer.eventLoop().execute(new Runnable() {
+                // This value may change, and so we should save it before executing the Runnable.
+                final boolean peerWriteInProgress = peer.writeInProgress;
+                peer.eventLoop().execute(new OneTimeTask() {
```

Comment:
The rational is because it would likely be inlined. Here is just a demo from running the HTTP/2 unit tests:

```
\-> TypeProfile (96801/96879 counts) = io/netty/channel/local/LocalChannel$LocalUnsafe
                                              @ 4   io.netty.channel.AbstractChannel::eventLoop (21 bytes)   inline (hot)
```